### PR TITLE
Implemented BrightnessAugmentation.java

### PR DESCRIPTION
**Pull request by sam-wmd**
Implemented BrightnessAugmentation class and BrightnessAugmentationTest classes

### CODE REVIEW

File: lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        assertTrue(Files.exists(outputPath), "Output image file should exist.");
+        assertTrue(Files.size(outputPath) > 0, "Output image file should not be empty.");
+    }
+    @Test
+    public void shouldIncreaseBrightness() throws IOException, InterruptedException {
+        // load test image
+        originalImage = loadTestImage("augmentation" + File.separator + "human-realistic.png");
+
+        // random double between 0 and 1
+        double rand = Math.random();
```
Comment (Samyssmile) [inline lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java:65]: replace with a value

File: lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        if (brightness < -1 || brightness > 1)
+            throw new IllegalArgumentException("Brightness value must be between -1 and 1");
+
+        this.brightnessFactor = brightness;
+    }
+
+    /**
+     * Loop through each image pixel and multiply the pixel value by the brightness value
+
+     * @param image The image to augment.
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java:35]: the

File: lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * limits the pixel value to a number between 0 and 255
+     * if pixelValue is less than 0, returns 0
+     * if pixelValue is greater than 255, returns 255
+     * else returns the original pixelValue
+     *
+     * @param pixelValue the value to be bound
+     * @return int
+     */
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java:68]: No JavaDoc for private elements, remove pls

File: lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        BrightnessAugmentation augmentation = new BrightnessAugmentation(rand);
+        augmentedImage = augmentation.apply(originalImage);
+
+        // check for each pixel if augmentedImage value is greater than original image value
```
Comment (Samyssmile) [inline lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java:69]: remove comment

File: lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                assertTrue(originalImage.getRGB(x,y) <= augmentedImage.getRGB(x,y));
+            }
+        }
+
+        openImageInPreview(originalImage);
+        openImageInPreview(augmentedImage);
+    }
+    @Test
+    public void shouldDecreaseBrightness() throws IOException, InterruptedException {
+        // load test image
```
Comment (Samyssmile) [inline lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java:81]: remove all comments

File: lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        originalImage = loadTestImage("augmentation" + File.separator + "national-park.png");
+
+        // random double between -1 and 0
+        double rand = Math.random() -1;
```
Comment (Samyssmile) [inline lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java:85]: replace with a value e.g. -0.5

File: lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        BrightnessAugmentation augmentation = new BrightnessAugmentation(rand);
+        augmentedImage = augmentation.apply(originalImage);
+        // check for each pixel if augmentedImage value is lower than original image value
```
Comment (Samyssmile) [inline lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java:88]: remove

File: lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            for (int x = 0; x < originalImage.getWidth(); x++){
+                assertTrue(originalImage.getRGB(x,y) >= augmentedImage.getRGB(x,y));
+            }
+        }
+
+        openImageInPreview(originalImage);
+        openImageInPreview(augmentedImage);
+    }
+
+
```
Comment (Samyssmile) [inline lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java:99]: Format Code. We use google formatte for java

File: lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import static de.edux.augmentation.AugmentationTestUtils.loadTestImage;
+import static de.edux.augmentation.AugmentationTestUtils.openImageInPreview;
+
+public class BrightnessAugmentationTest {
+
+    private BufferedImage originalImage;
+    private BufferedImage augmentedImage;
+
+
+    @AfterEach
```
Comment (Samyssmile) [inline lib/src/test/java/de/edux/augmentation/effects/BrightnessAugmentationTest.java:25]: nice approach

**Review by Samyssmile — CHANGES_REQUESTED** [review]

File: lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    @Override
+    public BufferedImage apply(BufferedImage image) {
+        BufferedImage augmentedImage =
+                new BufferedImage(image.getWidth(), image.getHeight(), image.getType());
+
+        for (int y = 0; y < image.getHeight(); y++){
+            for (int x = 0; x< image.getWidth(); x++){
+                int rgb = image.getRGB(x,y);
+                Color color = new Color(rgb);
+                int newR = clip((int) (color.getRed()   + 255f * brightnessFactor));
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java:47]: This looks wrong for me. getRed + x can bring you out rgb range. I think you need a proportial calculation here.

**Review by Samyssmile — CHANGES_REQUESTED** [review]

File: lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+                augmentedImage.setRGB(x,y,newColor.getRGB());
+            }
+        }
+
+        return augmentedImage;
+    }
+
+
+    /**
+     * limits the pixel value to a number between 0 and 255
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java:61]: Limits

File: lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    @Override
+    public BufferedImage apply(BufferedImage image) {
+        BufferedImage augmentedImage =
+                new BufferedImage(image.getWidth(), image.getHeight(), image.getType());
+
+        for (int y = 0; y < image.getHeight(); y++){
+            for (int x = 0; x< image.getWidth(); x++){
+                int rgb = image.getRGB(x,y);
+                Color color = new Color(rgb);
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java:46]: Let say we work with 1000x1000 Images. 1.000.000 Pixel there. You create 1.000.000 times new object here (2x 1.000.000) in this for loop. This is a critical thing here.

Instead I suggest working with primitives here.

`          int rgb = image.getRGB(x, y);
            int alpha = (rgb >> 24) & 0xff;
            int red = (rgb >> 16) & 0xff;
            int green = (rgb >> 8) & 0xff;
            int blue = rgb & 0xff;`

File: lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+                int newR = clip((int) (color.getRed()   + 255f * brightnessFactor));
+                int newG = clip((int) (color.getGreen() + 255f * brightnessFactor));
+                int newB = clip((int) (color.getBlue()  + 255f * brightnessFactor));
+
+                Color newColor = new Color(newR,newG,newB);
```
Comment (Samyssmile) [inline lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java:51]: Same here, do not create Color Objects for each pixel.
Do

`        int newRgb = (alpha << 24) | (red << 16) | (green << 8) | blue;`

**Review by Samyssmile — CHANGES_REQUESTED** [review]

File: lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java (hunk 10 above)
Comment (sam-wmd) [reply lib/src/main/java/de/edux/augmentation/effects/BrightnessAugmentation.java]: the clip() method is used here to keep the value within the RGB range. So the end result will always be coherent

**Review by Samyssmile — APPROVED** [review]

**Pull request merged**